prov/gni/src/gnix_sep.c

```diff
@@ -87,9 +89,31 @@ static void __trx_destruct(void *obj)
 	sep_priv = trx->sep;
 	assert(sep_priv != NULL);
 
+	/* These assignments must be done before we free ep_priv */
+	av = ep_priv->av;
+	index = ep_priv->src_addr.gnix_addr.cdm_id -
+		sep_priv->cdm_id_base;
+
 	refs_held = _gnix_ref_put(ep_priv);
-	if (refs_held == 0)
+
+	if (refs_held == 0) {
 		_gnix_ref_put(sep_priv->cm_nic);
+		/* Remove the context from the sep's list */
+		if ((index >= 0) &&
+		    (index < sep_priv->info->ep_attr->tx_ctx_cnt)) {
+			if (sep_priv->ep_table[index]) {
+				sep_priv->ep_table[index] = NULL;
+				_gnix_ref_put(av);
+			} else {
+				GNIX_WARN(FI_LOG_EP_CTRL,
```

Comment:
`refs_held` should only be zero once, so I think this works.  The __ref_put macro asserts if the ref count goes below zero.